kong/tools/request_aware_table.lua

```diff
@@ -0,0 +1,72 @@
+--- NOTE: tool is designed to assist with **detecting** request contamination
+-- issues on CI, during test runs. It does not offer security safeguards.
+
+local table_new = require "table.new"
+local base = require "resty.core.base"
+
+local debug_mode  = kong.configuration.log_level == "debug"
+local get_request = base.get_request
+local var         = ngx.var
+
+
+-- Check if access is allowed for table, based on the request ID
+local function enforce_sequential_access(table)
+  if next(table.__data) == nil or not get_request() then
+    -- empty table or not in a request context: allow access
+    rawset(table, "__allowed_request_id", nil)
+    return
+  end
+
+  local curr_request_id = var.request_id
+  local allowed_request_id = rawget(table, "__allowed_request_id")
+  if not allowed_request_id then
+    -- first access. Set allowed request ID and allow access
+    rawset(table, "__allowed_request_id", curr_request_id)
+    return
+  end
+
+  if curr_request_id ~= table.__allowed_request_id then
+    error("race condition detected; access to table forbidden", 2)
+  end
+end
+
+
+local _proxy_mt = {
+  __index = function(t, k)
+    enforce_sequential_access(t)
+
+    return t.__data[k]
+  end,
+
+  __newindex = function(t, k, v)
+    enforce_sequential_access(t)
+
+    t.__data[k] = v
+  end,
+}
+
+
+-- Request aware table constructor
+--
+-- Creates a new table with request-aware access logic to protect the
+-- underlying data from race conditions.
+--
+-- @param narr (optional) pre allocated array elements
+-- @param nrec (optional) pre allocated hash elements
+-- @return The newly created table with request-aware access
+local function new(narr, nrec)
+  if not narr then narr = 0 end
+  if not nrec then nrec = 0 end
+
+  local data = table_new(narr, nrec)
+  if debug_mode == false then
+    -- no patching if test mode is disabled
+    return data
+  end
+
+  return setmetatable({ __data = data }, _proxy_mt)
+end
+
+return {
+  new = new,
```

Comment:
Do we need a `:clear` method as well? How to clear `__` data?

Reply:
I see what you are saying. We can do this:

With debug mode on:
Use the proxy and metatable.

With debug mode on:
Use a different metatable `_mt_direct`, no proxy in this case. Define the `:clear` on this metatable only, not table instance.

This avoids you having to hack the "clear" method on the table each time, and prevents it being accidentally cleared by the user using `table.clear` manually.

Reply:
ok, I like that.